Corgibytes.Freshli.Cli/Functionality/Git/ComputeHistory.cs

```diff
@@ -9,48 +8,109 @@ namespace Corgibytes.Freshli.Cli.Functionality.Git;
+
+        if (rangeStartDate != startAtDate)
+        {
+            range.Add(rangeStartDate);
+        }
+
+        while (rangeStartDate > oldestCommit.CommittedAt)
+        {
+            var dateTimeOffset = quantifier switch
+            {
+                "d" =>
+                    // Go back at interval -x days
+                    rangeStartDate.AddDays(-interval),
+                "w" =>
+                    // Go back at interval of -7 * interval days
+                    rangeStartDate.AddDays(-7 * interval),
+                "m" =>
+                    // Go back at interval of -x months
+                    rangeStartDate.AddMonths(-interval),
+                "y" =>
+                    // Go back at interval of -x years
+                    rangeStartDate.AddYears(-interval),
+                _ => default
+            };
+
+            if (dateTimeOffset < oldestCommit.CommittedAt)
+            {
+                range.Add(oldestCommit.CommittedAt);
+                break;
+            }
+
+            range.Add(dateTimeOffset);
+            rangeStartDate = dateTimeOffset;
+        }
+
+        // Foreach offset in range, select the youngest commit, as long as it's not younger than the offset.
         return (
-                from groupedHistory in groupedHistories
-                select groupedHistory.MaxBy(i => i.CommittedAt)
-                into mostRecentCommitForThisGroupedPeriod
-                where mostRecentCommitForThisGroupedPeriod != null
-                select new HistoryIntervalStop(mostRecentCommitForThisGroupedPeriod.ShaIdentifier,
-                    mostRecentCommitForThisGroupedPeriod.CommittedAt))
-            .ToList();
+            from offset in range
+            let lastCommitForOffset = gitCommits.First(commit => commit.CommittedAt <= offset)
+            select new HistoryIntervalStop(lastCommitForOffset.ShaIdentifier, offset))
+        .ToList();
+    }
+
+    private static DateTimeOffset DetermineRangeStartDate(DateTimeOffset startAtDate, string? quantifier)
+    {
+        var rangeStartDate = startAtDate;
+        switch (quantifier)
+        {
+            case "w":
+                {
+                    // Start at first monday
+                    var daysToSubtract = (7 + (int)rangeStartDate.DayOfWeek - (int)DayOfWeek.Monday) % 7;
+                    rangeStartDate = rangeStartDate.AddDays(-daysToSubtract);
+                    break;
+                }
+            case "m":
+                // Start at first day of the month
+                rangeStartDate = new DateTimeOffset(rangeStartDate.Year, rangeStartDate.Month, 1, 0, 0, 0, rangeStartDate.Offset);
+                break;
+            case "y":
+                // Start at first day of the year
+                rangeStartDate = new DateTimeOffset(rangeStartDate.Year, 1, 1, 0, 0, 0, rangeStartDate.Offset);
+                break;
+        }
+
+        return rangeStartDate;
     }
```

Comment:
I am not understanding why the range start date is being forced to a Monday, the first day of the Month, or the first day of the year.  I was thinking that the interval parser was going to handle arbitrary date ranges.

Reply:
This logic is inherited from `freshli-lib`. You start on the current day, go to the start of the interval that the current day is in the middle of, and then skip backwards in time every X (days, weeks, months, or years depending on the selection). `freshli-lib` only supported an interval jump of 1 month, but this is the way that it handled it. This was needed for the month logic because a month is not a fixed length of time (not every month has a 31st day, but every month has a 1st day). In addition to the skips, the date of the oldest commit is always included, and the current date is always included. This means that the interval between the first and last pairs of data is not the same as the interval between the rest of the pairs of data.